### PR TITLE
A new dataset and a recipe 

### PR DESCRIPTION
**Pull request by JorisCos**
### **Content**
Add a new dataset derived from LibriSpeech ASR corpus http://www.openslr.org/12/
The recipe uses ConvTasNet with the MSS loss as a loss function

### CODE REVIEW

File: egs/libri_2_mix/ConvTasNet/run.sh [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+set -e  # Exit on error
+# Main storage directory. You'll need disk space to dump the raw LibriSpeech flac files
+# and generate the sources and mixtures
+storage_dir=/srv/storage/talc@talc-data.nancy/multispeech/calcul/users/jcosentino/DATA/
+
+# If you haven't downloaded and extracted the LibriSpeech dataset start from stage 1
+
+# If you already have dowloaded and extracted LibriSpeech,
+# specify the path to the directory here and start from stage 2
+LibriSpeech_dir=${storage_dir}
```
Comment (mpariente) [inline egs/libri_2_mix/ConvTasNet/run.sh]: Create a Librispeech directory inside storage_dir (at stage 1), do all the downloads there. LibriSpeech_dir would default to ${storage_dir}/LibriSpeech. 
Also, no caps in variables please : LibriSpeech_dir --> librispeech_dir

File: egs/libri_2_mix/ConvTasNet/local/prepare_data.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+. ../utils/parse_options.sh
+
+## Download WHAM noises
+mkdir -p $out_dir
+mkdir -p $storage_dir
+echo "Download LibiSpeech_train_100_hours_clean dataset into $storage_dir"
+# If downloading stalls for more than 20s, relaunch from previous state.
+wget -c --tries=0 --read-timeout=20 http://www.openslr.org/resources/12/train-clean-360.tar.gz -P $storage_dir
+tar -xzvf $storage_dir/train-clean-100.tar.gz -C $storage_dir
```
Comment (mpariente) [inline egs/libri_2_mix/ConvTasNet/local/prepare_data.sh]: Download 360 and untar 100, fix to 360.

File: egs/libri_2_mix/ConvTasNet/local/create_dataset_from_metadata.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+import pyloudnorm as pyln
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--librispeech_root_path', type=str, default=None,
+                    help='Path to librispeech root directory')
+parser.add_argument('--dataset_root_path', type=str, default=None,
+                    help='Path to the desired dataset root directory')
+parser.add_argument('--freqs', nargs='+', default=['8K'],
+                    help='--freqs 16K 8K will create 2 directories 8K '
+                         'and 16K')
```
Comment (mpariente) [inline egs/libri_2_mix/ConvTasNet/local/create_dataset_from_metadata.py]: Make the argument case-insensitive, maybe in `main()`, call upper() or lower().

File: asteroid/data/librimix_dataset.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        for i in range(self.n_src):
+            source_path = row[f'source_{i+1}_path']
+            s, _ = sf.read(source_path, dtype='float32')
+            if self.desired_length is not None:
+                s = s[:self.desired_frames]
+            sources_list.append(s)
+
+        mixture, _ = sf.read(mixture_path, dtype='float32')
+        if self.desired_length is not None:
+            mixture = mixture[:self.desired_frames]
```
Comment (mpariente) [inline asteroid/data/librimix_dataset.py:49]: I rather have a random start each time if the desired_length is specified. 
Rather than always starting from the beginning.

File: asteroid/data/librimix_dataset.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    Args:
+        metadata_path (str): The path to the metatdata file in
+        dataset/metadata
+        desired_length (int) : The desired sources and mixtures length in s
+        sample_rate (int) : The sample rate of the sources and mixtures
+        n_src (int) : The number of sources in the mixture
+    """
+
+    def __init__(self, metadata_path, desired_length=None, sample_rate=16000,
```
Comment (mpariente) [inline asteroid/data/librimix_dataset.py:19]: Let's keep the same interface a the other datasets and have a `segment` argument instead of `desired_length`.

File: asteroid/data/librimix_dataset.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                 n_src=2):
+        self.metadata_path = metadata_path
+        self.desired_length = desired_length
+        self.sample_rate = sample_rate
+        # Open csv file
+        self.df = pd.read_csv(metadata_path)
+        if self.desired_length is not None:
+            self.desired_frames = self.desired_length * self.sample_rate
```
Comment (mpariente) [inline asteroid/data/librimix_dataset.py:27]: This is a number of samples, not of frames.

File: asteroid/data/librimix_dataset.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        sources_list = []
+        # Read sources
+        for source_ID in source_ID_list:
+            s, _ = sf.read(os.path.join(self.sources_directory_path,
+                                        source_ID + '.wav'))
+            sources_list.append(s)
+
+        mixture, _ = sf.read(mixture_path)
+
+        mixture = torch.from_numpy(mixture).type(
```
[the remark this answers is not in the document]
Comment (mpariente) [reply asteroid/data/librimix_dataset.py:51]: What is the role of the `type`?

File: egs/libri_2_mix/ConvTasNet/eval.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
+import argparse
```
Comment (mpariente) [inline egs/libri_2_mix/ConvTasNet/eval.py:1]: The template eval script has changed a little bit, could you adapt to the newer versions please?

File: egs/libri_2_mix/ConvTasNet/local/conf.yml [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  early_stop: yes
+# Optim config
+optim:
+  optimizer: adam
+  lr: 0.001
+  weight_decay: 0.
+# Data config
+data:
+  metadata_train_path : "/home/jcosentino/libri/libri2mix/8K/min/metadata/mixture_train-clean-100.csv"
+  metadata_val_path : "/home/jcosentino/libri/libri2mix/8K/min/metadata/mixture_dev-clean.csv"
```
Comment (mpariente) [inline egs/libri_2_mix/ConvTasNet/local/conf.yml:34]: We're trying that only `run.sh` needs to be modified when running the recipe. We could create a local `./data` folder and copy things there? 
Let's try to avoid needing modifs in the conf file

File: egs/libri_2_mix/ConvTasNet/local/conf.yml [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  desired_length: 4
```
Comment (mpariente) [inline egs/libri_2_mix/ConvTasNet/local/conf.yml:35]: Can we change all the segment like the other recipes? 
If the name should be changed, we can make another PR about it.

File: egs/libri_2_mix/ConvTasNet/local/create_dataset_from_metadata.py [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+
+
+def main(arguments):
+    # Get librispeech root path
+    librispeech_root_path = arguments.librispeech_root_path
+    # Get dataset root path
+    dataset_root_path = arguments.dataset_root_path
+    # Get the desired frequencies
+    freqs = arguments.freqs
+    freqs = [freq.upper() for freq in freqs]
```
Comment (mpariente) [inline egs/libri_2_mix/ConvTasNet/local/create_dataset_from_metadata.py:31]: `freqs = [freq.upper() for freq in arguments.freqs]`

File: egs/libri_2_mix/ConvTasNet/local/create_dataset_from_metadata.py [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+    # Get the desired modes
+    modes = arguments.modes
+    modes = [mode.lower() for mode in modes]
```
Comment (mpariente) [inline egs/libri_2_mix/ConvTasNet/local/create_dataset_from_metadata.py:34]: Same here.

File: egs/libri_2_mix/ConvTasNet/local/create_metadata.py [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+    # If you wish to ignore some metadata files add their name here
+    # Example : to_be_ignored = ['dev-other.csv']
+    to_be_ignored = []
+    # Check if the metadata files in LibriSpeech already have been used
+    already_generated = os.listdir(mixtures_metadata_directory_path)
+    for generated in already_generated:
+        if generated.startswith("generating_mixture"):
+            to_be_ignored.append(generated.replace("generating_mixture_", ""))
+            print(f"{generated} already exists in "
+                  f"mixtures_metadata_directory_path it won't be overwritten")
```
Comment (mpariente) [inline egs/libri_2_mix/ConvTasNet/local/create_metadata.py:198]: Missing `{}`?

File: egs/libri_2_mix/ConvTasNet/model.py [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        est_masks = self.masker(tf_rep)
+        masked_tf_rep = est_masks * tf_rep.unsqueeze(1)
+        return self.pad_output_to_inp(self.decoder(masked_tf_rep), x)
+
+    @staticmethod
+    def pad_output_to_inp(output, inp):
+        """ Pad first argument to have same size as second argument"""
+        inp_len = inp.size(-1)
+        output_len = output.size(-1)
+        return nn.functional.pad(output, [0, inp_len - output_len])
```
Comment (mpariente) [inline egs/libri_2_mix/ConvTasNet/model.py:32]: This also changed since then.

File: egs/librimix/ConvTasNet/eval.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+    model = load_state_dict_in(state_copy, model)
+
+    # Handle device placement
+    if conf['use_gpu']:
+        model.cuda()
+    model_device = next(model.parameters()).device
+
+    test_set = LibriMix(conf['test_dir'], None,
+                        conf['sample_rate'],
+                        conf['train_conf']['data']['n_src'])
```
Comment (mpariente) [inline egs/librimix/ConvTasNet/eval.py:59]: Can you add the keywords in the call please? It makes it much easier to read. In particular, it's not obvious that `None` means, use all segment length, which is crucial !

File: asteroid/data/librimix_dataset.py [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+    Args:
+        metadata_path (str): The path to the metatdata file
+        sample_rate (int) : The sample rate of the sources and mixtures
+        n_src (int) : The number of sources in the mixture
+        segment (int) : The desired sources and mixtures length in s
+    """
+
+    def __init__(self, metadata_path, sample_rate=16000,
+                 n_src=2, segment='full'):
```
Comment (mpariente) [inline asteroid/data/librimix_dataset.py:20]: Can we keep the same behaviour as the other loaders : `segment=None` means full utterances

File: egs/librimix/ConvTasNet/eval.py [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+
+        mix_np = mix.cpu().data.numpy()
+        sources_np = sources.squeeze().cpu().data.numpy()
+        est_sources_np = reordered_sources.squeeze().cpu().data.numpy()
+        est_sources_np[0, :] = est_sources_np[0, :] / \
+                               np.max(est_sources_np, axis=1)[0] * \
+                               np.max(sources_np, axis=1)[0]
+        est_sources_np[1, :] = est_sources_np[1, :] / \
+                               np.max(est_sources_np, axis=1)[1] * \
+                               np.max(sources_np, axis=1)[1]
```
Comment (mpariente) [inline egs/librimix/ConvTasNet/eval.py:87]: We'll need to get rid of this. Estimate normalization cannot depend directly on the true source.

File: egs/librimix/ConvTasNet/local/conf.yml [hunk 18]
```diff
@@ -0,0 +1,37 @@
+encoder:
+  enc_name : free
+decoder:
+  dec_name : free
+
```
Comment (mpariente) [inline egs/librimix/ConvTasNet/local/conf.yml:5]: How about sticking to ConvTasnet for a baseline recipe? So we assume encoder/decoder use free filterbanks.

File: egs/librimix/ConvTasNet/local/conf.yml [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+  early_stop: yes
+# Optim config
+optim:
+  optimizer: adam
+  lr: 0.001
+  weight_decay: 0.
+# Data config
+data:
+  metadata_train_path : "/home/jcosentino/jcosentino/libri/small_snr/8K/min/metadata/mixture_train-clean-360.csv"
+  metadata_val_path : "/home/jcosentino/jcosentino/libri/small_snr/8K/min/metadata/mixture_dev-clean.csv"
```
Comment (mpariente) [inline egs/librimix/ConvTasNet/local/conf.yml:34]: We need a local path here, not an absolute one.
So the csv files need to be stored in data/wav8k/min/tr for example

File: egs/librimix/ConvTasNet/local/prepare_data.sh [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+tar -xzf $storage_dir/test-clean.tar.gz -C $storage_dir
+rm -rf $storage_dir/test-clean.tar.gz
+
+echo "Download LibiSpeech/train-clean-100 into $storage_dir"
+# If downloading stalls for more than 20s, relaunch from previous state.
+wget -c --tries=0 --read-timeout=20 http://www.openslr.org/resources/12/train-clean-100.tar.gz -P $storage_dir
+tar -xzf $storage_dir/train-clean-100.tar.gz -C $storage_dir
+rm -rf $storage_dir/train-clean-100.tar.gz
+
+echo "Download LibiSpeech/train-clean-360 into storage_dir"
```
Comment (mpariente) [inline egs/librimix/ConvTasNet/local/prepare_data.sh:28]: Miss the `$` to storage_dir to print it.

File: egs/librimix/ConvTasNet/train.py [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+def main(conf):
+    train_set = LibriMix(conf['data']['metadata_train_path'],
+                         conf['data']['sample_rate'],
+                         conf['data']['n_src'],
+                         conf['data']['segment'])
+
+    val_set = LibriMix(conf['data']['metadata_val_path'],
+                       conf['data']['sample_rate'],
+                       conf['data']['n_src'],
+                       conf['data']['segment'])
```
Comment (mpariente) [inline egs/librimix/ConvTasNet/train.py:38]: Same here, please us keyword arguments.

File: egs/librimix/ConvTasNet/eval.py (hunk 15 above)
Comment (JorisCos) [reply egs/librimix/ConvTasNet/eval.py:59]: Sure and you made me realize that my arguments don't match the dataloader arguments ...

File: asteroid/data/librimix_dataset.py (hunk 16 above)
Comment (JorisCos) [reply asteroid/data/librimix_dataset.py:20]: Ok, but I had a problem with that when I tried to set `segment =None' in the conf file when you load the .yml  None becomes a str "None" and the dataloader breaks. How do you think we should deal with that ?

File: egs/librimix/ConvTasNet/local/conf.yml (hunk 18 above)
Comment (JorisCos) [reply egs/librimix/ConvTasNet/local/conf.yml:5]: Sure

File: egs/librimix/ConvTasNet/eval.py (hunk 17 above)
Comment (JorisCos) [reply egs/librimix/ConvTasNet/eval.py:87]: Yeah that's bad. How do you usually deal with normalization ? I guess I could just avoid the clipping by dividing by 'max(est_source)'. Is relu better than sigmoid for mask activation ?

File: egs/librimix/ConvTasNet/local/conf.yml (hunk 19 above)
Comment (JorisCos) [reply egs/librimix/ConvTasNet/local/conf.yml:34]: Sure, but can you explain why please ?

File: asteroid/data/librimix_dataset.py (hunk 16 above)
Comment (mpariente) [reply asteroid/data/librimix_dataset.py:20]: I think `null` is what you're looking for.

File: egs/librimix/ConvTasNet/eval.py (hunk 17 above)
Comment (mpariente) [reply egs/librimix/ConvTasNet/eval.py:87]: This is one way, `src * max(abs(mix))/max(abs(src))` is also used but is not perfect. We could think about better. 
For the mask activation, I remember seeing better results for ReLU but cannot remember where. But, if better, it is marginal.

File: egs/librimix/ConvTasNet/local/conf.yml (hunk 19 above)
Comment (mpariente) [reply egs/librimix/ConvTasNet/local/conf.yml:34]: The user will modify one thing before running the recipe: `storage_dir` in run.sh. 
After that, the .csv files are stored in ./data and this path is given to the training script. 
This is one less thing for the user to do, so it is better.
If not stored, maybe symlinked.

**Comment by JorisCos** [comment]
I made a mistake when I pushed my last changes. The PR is now really messy. I will close this one and open new one
